### PR TITLE
[ragged-paged-attn] Use hidden states in kv cache and support any num_kv_head

### PR DESCRIPTION
**Pull request by bythew3i**
This PR uses hidden states (num_kv_head * head_dim) in kv cache. This change can unblock us with any num_kv_head. Previous if num_kv_head == 1 and dtype=bfloat16, we will have implicit padding in TPU. But now, after just using hidden states directly from projection, we no-longer need to use strided load, but just load by slice directly. 

This PR should help us support multi-chip sharding which shard num_kv_head to 1 for llama-3-70B.

Tested:
```
python test/test_pallas.py -v -k PallasTest.test_ragged_paged_attention_wrapper
```

### CODE REVIEW

File: torch_xla/experimental/custom_kernel.py [hunk 1]
```diff
@@ -1014,8 +955,7 @@ def ragged_paged_attention(
 ):
   if mask_value is None:
     mask_value = DEFAULT_MASK_VALUE
-  validate_ragged_paged_attention_inputs(q, k_pages, v_pages, kv_lens,
-                                         page_indices, cu_q_lens, num_seqs)
+
```
Comment (vanbasten23) [inline torch_xla/experimental/custom_kernel.py:958]: why stopped checking validate_ragged_paged_attention_inputs?

Comment (bythew3i) [reply torch_xla/experimental/custom_kernel.py:958]: Because we have these static shape check in JAX already

File: torch_xla/experimental/custom_kernel.py [hunk 2]
```diff
@@ -965,8 +905,9 @@ def _ragged_paged_attention_nonkernel(
     sm_scale=1.0,
     mask_value=DEFAULT_MASK_VALUE,
 ):
-  _, _, num_kv_heads, head_dim = k_pages.shape
-  num_q_heads = queries.shape[1]
+  _, num_q_heads, head_dim = queries.shape
+  _, _, kv_model_dim = k_pages.shape
```
Comment (vanbasten23) [inline torch_xla/experimental/custom_kernel.py:909]: nit: kv_hidden_size is more commonly used than kv_model_dim?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 3]
```diff
@@ -596,16 +593,19 @@ def ragged_paged_attention(
   Returns:
     The output of the attention.
   """
-  check_inputs_shapes(q, k_pages, v_pages, kv_lens, page_indices, cu_q_lens)
+  check_inputs_shapes(q, k_pages, v_pages, kv_lens, page_indices, cu_q_lens,
+                      num_seqs)
   num_q, num_q_heads, head_dim = q.shape
-  _, page_size, num_kv_heads, _ = k_pages.shape
+  _, page_size, kv_model_dim = k_pages.shape
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:599]: nit: kv_hidden_size is more commonly used than kv_model_dim?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 4]
```diff
@@ -520,31 +527,22 @@ def get_dtype_packing(dtype):
...
-  min_q_heads = min_kv_heads * ratio
-  if can_be_xla_fully_tiled(min_q_heads, q_packing):
-    return min_q_heads, min_kv_heads
-  return num_q_heads, num_kv_heads
+  def lcm(a, b):
+    return a * b // gcd(a, b)
+
+  q_packing = get_dtype_packing(q_dtype)
+  max_q_tiling = 8 * q_packing
+  min_q_heads = lcm(max_q_tiling, num_q_heads_per_kv_head)
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:542]: I am not sure if I follow. If dtype is bf16, then max_q_tiling is 16. If it's qwen where num_q_heads=12, kum_kv_head=2, num_q_heads_per_kv_head=6, then min_q_heads (=lcm(max_q_tiling, num_q_heads_per_kv_head)) will be 48. What does `min_q_heads` mean?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 5]
```diff
@@ -107,25 +120,26 @@ def validate_inputs_on_runtime(
...
   max_num_batched_tokens = q.shape[0]
   page_size = k_pages.shape[1]
   max_num_seqs, pages_per_seq = page_indices.shape
-  if num_seqs > max_num_seqs:
-    raise ValueError(f"{num_seqs=} must be less or equal to {max_num_seqs=}")
+  if num_seqs[0] > max_num_seqs:
+    raise ValueError(f"{num_seqs[0]=} must be less or equal to {max_num_seqs=}")
   max_kv_len = jnp.max(kv_lens)
-  min_pages_per_seq = ceil_div(max_kv_len, page_size)
+  min_pages_per_seq = cdiv(max_kv_len, page_size)
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:133]: why is it min? Shouldn't it be max_pages_per_seq since you used cdiv(jnp.max(kv_lens), page_size)?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 6]
```diff
@@ -136,36 +150,44 @@ def validate_inputs_on_runtime(
...
     raise ValueError(
-        f"{k_pages.shape=} and {v_pages.shape=} must have the same shape.")
-  if head_dim_k != head_dim:
-    raise ValueError(
-        f"Q head_dim {head_dim} must be the same as that of K/V {head_dim_k}.")
+        f"Expected {k_pages.shape=} to be equal to {v_pages.shape=}.")
+  _, page_size, kv_model_dim = k_pages.shape
+  kv_packing = get_dtype_packing(k_pages.dtype)
+  if page_size % kv_packing != 0:
+    raise ValueError(f"Expected {page_size=} is divisible by {kv_packing=}")
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:169]: page_size % kv_packing != 0 indicating there will be padding so we may waste some memory. Can we give a warning instead of raising an exception?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 5 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:133]: That is lower bound for `pages_per_seq`

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 6 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:169]: The page size is chosen by the serving config, the error indicates we should choose better one. Otherwise when people using bf16 or quantized types (fp8, int8, int4) there will be no bandwidth saving. We should prevent this.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 4 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:542]: It tries to find a min number that is fully divisible by either `max_q_tiling` or `num_q_heads_per_kv_head`, if this number can divide total num_q_heads evenly, we just use this number as num_q_heads_per_blk. If we can not find one, we use the total num_q_heads .

Checking if it is divisible by `max_q_tiling` is to make sure it can be fully tiled by XLA. 
Checking if it is divisible by `num_q_heads_per_kv_head` is to make sure we do not need to have inner split in `num_q_heads_per_kv_head`.

File: torch_xla/experimental/custom_kernel.py (hunk 2 above)
Comment (bythew3i) [reply torch_xla/experimental/custom_kernel.py:909]: Done.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 3 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py]: Done

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 6 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:169]: I see. I guess it's the same reason why before this PR when num_kv_head==1 and dtype=bf16, we would raise an exception https://github.com/pytorch/xla/blob/c7d0b1efb37c3bfcfc1cfa050930193234dd2aa5/torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py#L534-L536 (`Previous if num_kv_head == 1 and dtype=bfloat16, we will have implicit padding in TPU.`). The point is the code may still run fine but there will be no bandwidth savings.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 5 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:133]: gotcha

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 4 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:542]: Thanks! Could you add what you said as a comment in the code?

**Review by vanbasten23 — APPROVED** [review]
Thanks Jevin. LGTM pending on CI.

**Comment by vanbasten23** [comment]
@bythew3i I assume you have run the tests tests/pallas/tpu_ragged_paged_attention_test.py and they all pass?

**Comment by bythew3i** [comment]
> @bythew3i I assume you have run the tests tests/pallas/tpu_ragged_paged_attention_test.py and they all pass?

Yes I tested the kernel.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py (hunk 6 above)
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:169]: Yes, the point of quantization is to save more memory and bandwidth
